ansible_base/rbac/models.py

```diff
@@ -0,0 +1,397 @@
+    the data in this table is created from the ObjectRole and RoleDefinition data
+      you should not interact with this table yourself
+      the only method that should ever write to this table is
+        compute_object_role_permissions()
+
+    In the above example, "ObjectRole 423" may be a role that grants membership
+    to a team, and that team was given permission to another ObjectRole.
+    """
+
+    class Meta:
+        app_label = 'dab_rbac'
+        verbose_name_plural = _('role_object_permissions')
+        indexes = [
+            models.Index(fields=["role", "content_type_id", "object_id"]),  # used by get_roles_on_resource
+            models.Index(fields=["role", "content_type_id", "codename"]),  # used by accessible_objects
+        ]
+        constraints = [models.UniqueConstraint(name='one_entry_per_object_permission_and_role', fields=['object_id', 'content_type_id', 'codename', 'role'])]
+
+    def __str__(self):
+        return (
+            f'RoleEvaluation(pk={self.id}, codename={self.codename}, object_id={self.object_id}, '
+            f'content_type_id={self.content_type_id}, role_id={self.role_id})'
+        )
+
+    def save(self, *args, **kwargs):
+        if self.id:
+            raise RuntimeError('RoleEvaluation model is immutable and only used internally')
+        return super().save(*args, **kwargs)
+
+    role = models.ForeignKey(
+        ObjectRole, null=False, on_delete=models.CASCADE, related_name='permission_partials', help_text=_("The object role that grants this form of permission")
+    )
+    codename = models.TextField(null=False, help_text=_("The name of the permission, giving the action and the model, from the Django Permission model"))
+    # NOTE: we awkwardly do not form these into a content_object, following from AWX practice
+    # this can be relaxed as we have comparative performance testing to confirm doing so does not affect permissions
+    content_type_id = models.PositiveIntegerField(null=False)
+    object_id = models.PositiveIntegerField(null=False)
+
+    def obj_perm_id(self):
+        "Used for in-memory hashing of the type of object permission this represents"
+        return (self.codename, self.content_type_id, self.object_id)
+
+    @staticmethod
+    def accessible_ids(cls, actor, codename, content_types=None):
+        """
+        Corresponds to AWX accessible_pk_qs
+
+        Use instead of `MyModel.objects` when you want to only consider
+        resources that a user has specific permissions for. For example:
+        MyModel.accessible_objects(user, 'view_mymodel').filter(name__istartswith='bar')
+
+        Intended to be used for users, but should also be valid for teams
+        """
+        # We only have a content_types exception for multiple content types for polymorphic models
+        # for normal models you should not need it, but AWX unified_ models need it to get by
+        filter_kwargs = dict(role__in=actor.has_roles.all(), codename=codename)
```

Comment:
Could you do this instead?

```
filter_kwargs = dict(role__in=actor.has_roles.filter(role_definition__permissions__codename=codename))
```

This would make it so you don't have to add "codename" to the RoleEvaluation table. Would this be less efficient?

Reply:
There's an explicit index for `["role", "content_type_id", "codename"]`. It's true that all foreign keys are going to be an index by itself.

But the logic won't quite work. The `codename` here is to be the resultant object permission. You could be a have the member role for a team, which lists like 2 permissions, both for the team CT. What we need to know from this table is the fact that the team role gives update permission to inventory 52, for example. That data is written very verbosely in a row here.